dvc/utils/__init__.py

```diff
@@ -480,3 +481,48 @@ def glob_targets(targets, glob=True, recursive=True):
         for target in targets
         for exp_target in iglob(target, recursive=recursive)
     ]
+
+
+@contextmanager
+def intercept_error(onerror: Callable = None, **kwargs):
+    try:
+        yield
+    except Exception as e:  # pylint: disable=W0703
+        logger.debug("", exc_info=True)
+        if onerror is not None:
+            onerror(e, **kwargs)
+
+
+class Onerror:
```

Comment:
Would also like to mention that I think the naming convention here is a bit confusing.

In our methods where we want to pass the error handling callback, we define the param as `onerror=None`, where `onerror` should be a callback function that accepts 2 args (exception instance and revision) and one kwarg (optional str path). (This callback format isn't mentioned anywhere, but should probably be mentioned in the `Repo.func()` level docstrings)

This class is also named `Onerror`, but what it does is handle collecting any errors passed into it (and they get passed via the callback method which is defined in `__call__`). I think my main issue with the naming is that it should be something along the lines of `ErrorCollector`. 

The way it's named now, it looks like users using the `Repo` API are supposed to subclass `Onerror`, and pass those subclassed objects into `onerror=...`, when all they really need to do is implement a callback method. (Because not everyone will want to collect errors, in some cases I may just want to check for a specific exception type to `pass`, and re-raise everything else.)